### ISSUE TITLE
System.Posix.Types.Orphans.CDev, Bits instance, implements bit fails with GHC/LLVM

### ISSUE THREAD

**Issue by ip1981** [haskell-compat/base-orphans#47]
```
Failures:

  test/System/Posix/Types/OrphansSpec.hs:51:5: 
  1) System.Posix.Types.Orphans.CDev, Bits instance, implements bit
       Falsifiable (after 2 tests):
         -1
```

**Comment by RyanGlScott** [comment]
Odd, the tests are passing on Travis CI with GHC 8.6.3 on Linux (see [here](https://travis-ci.org/haskell-compat/base-orphans/jobs/465953279)). Are you using a different operating system by any chance? Do the tests work for you on other versions of GHC (e.g., 8.4.4 or 8.6.2)?

**Comment by ip1981** [comment]
Oh, I got it with LLVM backend (6.0.x) via `cabal new-build --enable-tests --ghc-options='-fllvm -O2'`.

I didn't verify it with older GHC. With the default backend (GCC?) there is no failure.

**Comment by RyanGlScott** [comment]
Hm, interesting. I think the fact that this doesn't work with `-fllvm` is due to a bug in the way that GHC optimizes this program:

```haskell
module Main (main) where

import Data.Bits
import Data.Word

main :: IO ()
main = print $ toInteger (shiftL 1 hm :: Word64)
            == toInteger (shiftL 1 hm :: Word64)

hm :: Int
hm = -1
{-# NOINLINE hm #-}
```

When compiled with `-O1`, this returns `True`, but when compiled with `-O1 -fllvm`, this returns `False`! I've filed this as [Trac #16111](https://ghc.haskell.org/trac/ghc/ticket/16111).

In any case, we probably shouldn't be passing negative numbers (like -1) as the second argument to `shiftL` anyway. I've pushed a commit to ensure that this doesn't happen, so this should be fixed on `base-orphans`' side in 8ff04538402c925d64b4634d7af8358428d1e9fb.
